### ISSUE TITLE
[PERF]: GPUs starved by the single-threaded input pipeline — audit first, then optimize the producer

### ISSUE THREAD

**Issue by zachtheyek** [zachtheyek/Ether-Scan#276]
## Summary

During the Phase-2 release run on bla0 (`train_20260724_215156`) the GPUs are **idle the large majority of every epoch**. Training is not compute-bound, disk-bound, or CPU-bound — it is bound by the host-side input path that feeds the GPUs. Epochs take ~13 min (bla0) / ~15 min (blpc3) where the recorded GPU benchmarks predict roughly **5× faster**.

This issue covers fixing that. **It must start with a wholesale performance audit** to confirm the exact bottleneck before any optimization lands — the measurements below are strong but they localize the problem, they do not yet prove the precise mechanism.

## Evidence (measured live on bla0, read-only, 2026-07-26)

| signal | observation | reading |
| --- | --- | --- |
| GPU utilization | `0%` on 12 of 15 one-second samples; one burst to 28–86% | starved, not saturated |
| VRAM | 4.5 GB used of 16 GB (bla0); 5 GB of 96 GB (blpc3) | enormous headroom |
| CPU | process at **144% of 9600%** (1.4 of 96 cores), 402 threads | not CPU-bound |
| Disk | `wa=0`, `bi=0`, 396 GB page cache vs a 275 GB round | **not disk-bound** — the round data is resident in RAM |
| Context switches | **~700,000/s** at 95% idle | coordination/latency overhead, not work |

Two independent cross-checks:

1. **blpc3 (5× RTX PRO 6000 96 GB) is *slower* (~15 min/epoch) than bla0 (6× A4000 16 GB, ~13 min/epoch).** If this were compute-bound that would be impossible.
2. The recorded benchmark for bla0 is 3,087 cadences/s on 6 GPUs (`benchmarks/README.md`, PR #253), so a 399,360-cadence epoch should be ~2.2 min of GPU work. Observed train phase is ~9–11 min ⇒ GPUs are doing real work roughly **20–25%** of the epoch.

## Prime suspect (to be confirmed by the audit, not assumed)

[`src/aetherscan/train.py:686`](../blob/master/src/aetherscan/train.py#L686):

```python
train_dataset = (
    tf.data.Dataset.from_generator(train_generator, output_signature=output_signature)
    .repeat()
    .prefetch(tf.data.AUTOTUNE)
)
```

The producer is a single Python generator thread — no `num_parallel_calls`, no `interleave`, no parallel `map`. `prefetch` overlaps *one* producer with the consumer; it does not parallelize production. Each global step that thread must gather ~151 MB (768 cadences × 6 × 16 × 512 × 4 B) via a fancy-index gather (`concat[batch_indices]`) out of a 275 GB disk-backed memmap, with a **global shuffle each epoch** (batch membership is random across the whole round, only sorted *within* a batch for locality), then hand it to a 6-GPU `experimental_distribute_dataset`.

`benchmarks/bench_gpu.py` drives synthetic in-memory tensors, so this path has **never** been benchmarked — `docs/BENCHMARKING.md` / the PR #253 notes explicitly list the input pipeline as an unmeasured gap, assumed low-value "given compute-bound". This run falsifies that assumption.

## Required first step — wholesale performance audit

Do **not** start editing the input pipeline. First measure, so we do not optimize the wrong thing:

- [ ] **TF Profiler** over a few steps (`tf.profiler.experimental`) → the authoritative host-vs-device timeline: `tf.data` input-pipeline analysis, step-time breakdown, host→device transfer time, per-op device time.
- [ ] **`tf.data` specifics**: is the bottleneck the generator itself, the `experimental_distribute_dataset` split/copy, or H2D? Use the profiler's input-pipeline analyzer plus `tf.data.experimental.StatsAggregator` if useful.
- [ ] **`py-spy dump` / `py-spy top`** on the training process → where the generator thread actually spends its time (page faults, numpy gather, queue waits, lock waits).
- [ ] **Attribute the ~700k ctx-switches/s** — this is unexplained and may be the real story (thread-pool spinning, NCCL/all-reduce sync, gradient-accumulation barriers).
- [ ] **Isolate `--tf-deterministic-ops`**: `tf.config.experimental.enable_op_determinism()` ([train.py:902](../blob/master/src/aetherscan/train.py#L902)) forces deterministic kernels **and** deterministic `tf.data` ordering — the latter directly blocks the parallel-map reordering any fix would rely on. Quantify its cost separately; it trades against bit-exact reproducibility, which is a maintainer call.
- [ ] **Isolate the background `RoundDataProducer`** — round N+1 generation runs concurrently with round N training and competes for the same I/O/page cache (especially under `--keep-round-data`). Measure with and without.
- [ ] Establish a **repeatable micro-benchmark** for the real input path (memmap → tf.data → distributed dataset → step) so any fix is provably better. This closes the `bench_gpu.py` gap and belongs in `benchmarks/`.

Only once the audit names the bottleneck should optimization land, with before/after numbers from that micro-benchmark.

## Candidate optimizations (hypotheses, pending the audit)

- Parallelize the gather: replace the single generator with an index-batch dataset + `.map(gather_fn, num_parallel_calls=AUTOTUNE)` (numpy releases the GIL during the memmap copy, so this should genuinely scale), or shard per replica.
- Trade the per-batch global random gather for block-sequential reads plus a shuffle buffer — far better locality for the same statistical effect.
- `tf.data.experimental.prefetch_to_device` / deeper prefetch so H2D overlaps compute.
- Revisit `--tf-deterministic-ops` for runs that don't need bit-exactness.

Explicitly **not** promising: bigger batches, XLA, or mixed precision. The GPUs are starved, and PR #253's sweep already found batch 128 throughput-optimal with spare VRAM unable to buy speed.

## Sequencing

Work starts **after the blpc3 run (`test_20260724_190816`) completes and its accidental HF tag is deleted** — blpc3 then becomes the test bed, so the bla0 release run is never disturbed. The bla0 run continues untouched at ~13 min/epoch (~8–9 days); whether it is worth restarting on a proven speedup is a separate maintainer decision once we have numbers.

## Acceptance criteria

- [ ] Audit report identifying the bottleneck with profiler evidence (committed under `benchmarks/`).
- [ ] A repeatable real-input-path benchmark.
- [ ] A fix with before/after throughput and GPU-utilization numbers.
- [ ] Bit-exactness/reproducibility implications stated explicitly if determinism settings change.
- [ ] No change to training semantics (same shuffling guarantees, same epoch composition) — or an explicit, argued case if that changes.

**Comment by zachtheyek** [comment]
Cross-link: #277 (DB writer backlog / plot correctness) has a **genuine overlap with this audit** worth sequencing around.

The injection-stat write path runs on **main-process threads under the GIL**: the `RoundDataDrainer` thread performs ~300 K `write_injection_stat()` calls per producer message, while the `tf.data` `from_generator` producer — the prime suspect for the starvation here — is *also* a main-process Python thread. GIL contention between them is a plausible contributor to the ~700 k context-switches/s at 95% idle measured on bla0.

**Implication for this issue's audit:** attribute **GIL / thread time explicitly** (e.g. `py-spy dump` across all threads, not just the generator), and measure with the producer drainer both active and idle. Otherwise #277's write-path fix and any input-pipeline change here will be conflated. Run this audit first, then re-measure GPU utilisation after #277's batching lands.
